File: packages/shared-types/src/market.ts
export interface RentData {
  id: string;
  county_id: number | null;
  zip_code: string | null;
  period_year: number;
  period_month: number;
  asking_rent_1br: number | null;
  asking_rent_2br: number | null;
  effective_rent_1br: number | null;
  effective_rent_2br: number | null;
  vacancy_rate: number | null;
  yoy_rent_growth_pct: number | null;
  data_source: string;
}

export interface DemographicData {
  id: string;
  county_id: number;
  acs_vintage_year: number;
  population_current: number | null;
  population_3yr_prior: number | null;
  population_growth_rate: number | null;
  renter_occupied_pct: number | null;
  median_household_income: number | null;
  median_age: number | null;
  bebr_projected_population: number | null;
}

export interface EmploymentData {
  id: string;
  county_id: number | null;
  msa_name: string | null;
  period_year: number;
  period_month: number;
  total_employment: number | null;
  yoy_employment_growth_pct: number | null;
  unemployment_rate: number | null;
  sector_employment: Record<string, number>;
}

export interface SupplyPipeline {
  id: string;
  county_id: number;
  period_year: number;
  period_month: number;
  units_under_construction: number | null;
  units_permitted_ltm: number | null;
  existing_inventory: number | null;
  supply_pressure_ratio: number | null;
  data_source: string;
}

export interface CountyMarketSnapshot {
  county_id: number;
  county_name: string;
  latest_rent: RentData | null;
  demographics: DemographicData | null;
  employment: EmploymentData | null;
  supply: SupplyPipeline | null;
}

File: apps/api/models/market.py
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class RentData(BaseModel):
    id: UUID
    county_id: Optional[int] = None
    zip_code: Optional[str] = None
    period_year: int
    period_month: int
    asking_rent_1br: Optional[float] = None
    asking_rent_2br: Optional[float] = None
    effective_rent_1br: Optional[float] = None
    effective_rent_2br: Optional[float] = None
    vacancy_rate: Optional[float] = None
    yoy_rent_growth_pct: Optional[float] = None
    data_source: str


class DemographicData(BaseModel):
    id: UUID
    county_id: int
    acs_vintage_year: int
    population_current: Optional[int] = None
    population_3yr_prior: Optional[int] = None
    population_growth_rate: Optional[float] = None
    renter_occupied_pct: Optional[float] = None
    median_household_income: Optional[float] = None
    median_age: Optional[float] = None
    bebr_projected_population: Optional[int] = None


class EmploymentData(BaseModel):
    id: UUID
    county_id: Optional[int] = None
    msa_name: Optional[str] = None
    period_year: int
    period_month: int
    total_employment: Optional[int] = None
    yoy_employment_growth_pct: Optional[float] = None
    unemployment_rate: Optional[float] = None
    sector_employment: dict = {}


class SupplyPipeline(BaseModel):
    id: UUID
    county_id: int
    period_year: int
    period_month: int
    units_under_construction: Optional[int] = None
    units_permitted_ltm: Optional[int] = None
    existing_inventory: Optional[int] = None
    supply_pressure_ratio: Optional[float] = None
    data_source: str


class CountyMarketSnapshot(BaseModel):
    county_id: int
    county_name: str
    latest_rent: Optional[RentData] = None
    demographics: Optional[DemographicData] = None
    employment: Optional[EmploymentData] = None
    supply: Optional[SupplyPipeline] = None
